### PR TITLE
log/file: Reduce log file flush frequency, if configured.

### PR DESCRIPTION
**Pull request by jlucovsky**
This PR provides a configurable way to reduce the number of times an output file is flushed.

By default, output files will continue to be flushed after each write. When a flush-threshold is configured, files will be flushed every `flush-threshold` bytes; the byte count since the last write is tracked and when it reaches or exceeds the threshold, the file will be flushed.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [3449](https://redmine.openinfosecfoundation.org/issues/3449)

Describe changes:
- Refactor write/clearerror logic to remove duplication
- Add `flush-threshold` config setting to control `fflush` frequency
- When a threshold exists, track flush count and display (as perf) when output file is closed.

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the pull request number.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

File: src/util-logopenfile.c [hunk 1]
```diff
@@ -517,6 +506,15 @@ SCConfLogOpenGeneric(ConfNode *conf,
     if (filetype == NULL)
         filetype = DEFAULT_LOG_FILETYPE;
 
+#define DEFAULT_LOG_FLUSH_THRESHOLD 0
```
Comment (jasonish) [inline src/util-logopenfile.c:509]: Not used.

File: doc/userguide/output/eve/eve-json-output.rst [hunk 2]
```diff
@@ -301,6 +307,32 @@ This example will cause each Suricata thread to write to its own "eve.json" file
...
+up to ``flush-threshold`` bytes unbuffered/unwritten should Suricata unexpectedly terminate.
+
+The default value is ``0``.
+
+::
+
+   outputs:
+     - eve-log:
+         filename: eve.json
+         flush-threshold: 1024
```
Comment (jasonish) [inline doc/userguide/output/eve/eve-json-output.rst:329]: As no alternatives to the default are provided, users may look at this example as a suggestion, however, given average log sizes a value of 1024 might not help much - might reduce the flush to every 2nd log record. Would 4k or 8k be more reasonable?

**Comment by victorjulien** [comment]
I guess what I don't like about such approaches if that if traffic suddenly stops (common in testing for example), the last logs may be held indefinitely. A time based approach might make more sense, but is probably more complex.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .app_layer.error.snmp.parser | 1 | 0 | - | 

Pipeline 13517

File: doc/userguide/output/eve/eve-json-output.rst (hunk 2 above)
Comment (jlucovsky) [reply doc/userguide/output/eve/eve-json-output.rst:329]: I'll update the example.

However, the timing issue (lingering data) is a problem that I have to solve if this PR is to move forward.

File: src/util-logopenfile.c (hunk 1 above)
Comment (jlucovsky) [reply src/util-logopenfile.c:509]: Will remove.

**Comment by victorjulien** [comment]
I've been thinking about time issues before, and been wondering if it would make sense to create "heartbeat" like pseudo packets that are flushed through on an interval of say 1 second. Then we'd be able to do housekeeping tasks, like flushing lingering log data.

**Comment by jlucovsky** [comment]
> I guess what I don't like about such approaches if that if traffic suddenly stops (common in testing for example), the last logs may be held indefinitely. A time based approach might make more sense, but is probably more complex.

The bytes are always written; this PR only reduces the calls to `fflush()`.

File: src/util-logopenfile.c [hunk 3]
```diff
@@ -240,43 +252,15 @@ static int SCLogFileWriteNoLock(const char *buffer, int buffer_len, LogFileCtx *
...
-            SCConfLogReopen(log_ctx);
-        }
-
-        if (log_ctx->flags & LOGFILE_ROTATE_INTERVAL) {
-            time_t now = time(NULL);
-            if (now >= log_ctx->rotate_time) {
-                SCConfLogReopen(log_ctx);
-                log_ctx->rotate_time = now + log_ctx->rotate_interval;
-            }
-        }
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/util-logopenfile.c:264]: It does not; these steps are performed in `SCLogFileWriteNoLock`.

**Comment by jlucovsky** [comment]
I plan to revisit this after the 7.0 release.

The main issue is how to handle periods of inactivity and at shutdown and ensure that data that's been written is actually flushed.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .app_layer.error.snmp.parser | 1 | 0 | - | 

Pipeline 13517

**Comment by jlucovsky** [comment]
Continued in #9587

**Pull request closed without merging**